### PR TITLE
[NO GBP] Final precision rounding for reagent volumes

### LINKED ISSUE #79566: Chem factories are bugged once again
Since their latest patch, the chemistry code and chem factories were finally working again, aside from the floating point errors with bottles ending in 49.99 units instead of 50.
I tried to make another one, and I've encountered some more errors.

This error is back with the reaction chambers, where chems will stop collecting inside them at .99 units, halting the mixing process. This seems to occur just at random, and will fix itself once plunged only to get jammed again after a cycle or two:
6ARKndn
FwIeEi6
NyR9DFB

Also, I'm not sure if this is an issue or a feature, but reaction chambers will now skip an ingredient if it doesn't have it and start collecting the others, when before it will pause at which ingredient is missing and wait for it before collecting the others:
o1TNdIH

@SyncIt21 I assume this is your issue

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
- Fixes #79566

This applies mostly to plumbing reaction chambers but to implement that fix some rounding operations had to be carried over to `holder.dm`(which will benefit everything in general)

I'm pulling out all the stops here. Rather than checking "are we close enough" plumbing reaction chambers will now check if we and i quote "absolutely insanely precisely there". This means volumes like 49.9999 should become 50 period.

Note this is a high probability & not a definite fix. i.e. now theoretically 100% of the time you should not get this problem but if it still happens then as of now i have no solution and have to go back to the drawing board on this one but i am very confident this should be the end of all plumbing related problems i.e. at least problems with volumes not getting rounded to whole numbers

## Changelog
:cl:
fix: plumbing factories should not rarely/randomly brick at volumes like 0.9999(when in fact it should have been 1)
/:cl:

### CODE REVIEW

**Comment by SyncIt21** [comment]
@MidoriWroth go ahead.

**Comment by MidoriWroth** [comment]
i'll give it a go when i get home from work in a bit

**Comment by SyncIt21** [comment]
> i'll give it a go when i get home from work in a bit

Yup. This will remain drafted till you say so

**Comment by MidoriWroth** [comment]
well i tested it and let the factory run for a while, seems to have no actual issues and works fine.

**Comment by MidoriWroth** [comment]
oh actually i noticed the new table salt recipe (sodium + chlorine) wasn't working, so i checked the reagent search and it was this: 
5ilvWyR

Is that supposed to be the new recipe now? seems kinda wild if so lmao

**Comment by CliffracerX** [comment]
> oh actually i noticed the new table salt recipe (sodium + chlorine) wasn't working, so i checked the reagent search and it was this: 5ilvWyR
> 
> Is that supposed to be the new recipe now? seems kinda wild if so lmao

During my testing I found that Sodium + Chlorine still works, this seems to be a bug with the reaction search helper.

**Comment by SyncIt21** [comment]
> well i tested it and let the factory run for a while, seems to have no actual issues and works fine.

Yeeeaah!!! Thank you for your patience. If you want me yo undraft it just give the word

**Comment by MidoriWroth** [comment]
yeah go ahead, seems fine now

**Comment by SyncIt21** [comment]
> oh actually i noticed the new table salt recipe (sodium + chlorine) wasn't working, so i checked the reagent search and it was this: 5ilvWyR
> 
> Is that supposed to be the new recipe now? seems kinda wild if so lmao

- #79499 

Did change some values. That or some other PR but not mine

**Comment by SyncIt21** [comment]
Anyway if you want to test some more go ahead or if you me to draft just let me know.

**Comment by SyncIt21** [comment]
> yeah go ahead, seems fine now

Yup time to undraft

File: code/modules/plumbing/plumbers/reaction_chamber.dm [hunk 1]
```diff
@@ -109,30 +109,32 @@
...
 
 			if(!required_reagents.Find(input_reagent))
 				var/input_amount = text2num(params["amount"])
-				if(input_amount)
+				if(!isnull(input_amount))
 					required_reagents[input_reagent] = input_amount
-
-			return TRUE
+					. = TRUE
+			return
```
Comment (Jacquerel) [inline code/modules/plumbing/plumbers/reaction_chamber.dm:123]:
```suggestion
					return TRUE
			return FALSE
```
If this is your intention then we should make it clear.

File: code/modules/plumbing/plumbers/reaction_chamber.dm [hunk 2]
```diff
@@ -109,30 +109,32 @@
...
 
 		if("remove")
 			var/reagent = get_chem_id(params["chem"])
 			if(reagent)
 				required_reagents.Remove(reagent)
-			return TRUE
+				. = TRUE
+			return
```
Comment (Jacquerel) [inline code/modules/plumbing/plumbers/reaction_chamber.dm:130]: Same in both following cases.

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by GPeckman** [comment]
I unironically think that it would be best to store reagent volume as an integer in the code (e.g. 1u to a player is stored as 100 internally, 0.01u is just 1) and avoid floating point nonsense altogether. Material code in protolathes and such already does this, and as far as I know there hasn't been any issues like this there.

**Comment by SyncIt21** [comment]
> I unironically think that it would be best to store reagent volume as an integer in the code (e.g. 1u to a player is stored as 100 internally, 0.01u is just 1) and avoid floating point nonsense altogether. Material code in protolathes and such already does this, and as far as I know there hasn't been any issues like this there.

That was my first thought however a lot of stuff in-game trickle in very small fractional values over time and if we round them to whole numbers then their effect can dissapear especially when they are very close to the target.

I don't want to take that risk so we'll see how far this goes. Material lathes and stuff don't have this problem because literally everything costs well about 1 unit however when chems react we can't round of that precision else we risk losing data so let's see where this takes us.

**Pull request merged**